posts/2018-12-11-hunt-bugs-down-before-they-are-merged.md

```diff
@@ -0,0 +1,362 @@
+
+```
+
+The following HTTP call returns the list of all of a user's notes:
+
+``` http
+GET /users/<user>/notes?query=<prefix>
+```
+</div>
+
+### It gets in the way
+
+<div class="story" style="display: inline-block;">
+
+<figure style="float: left;">
+<img src="../images/peanut_butter_cookies_nola.jpg" alt="Rated #1 in NOLA"><figcaption>And #2 in America!</figcaption>
+</figure>
+
+You've been doing amazing work. You got the flow. Mind like water. You're done
+with your frontend changes. You spin up a local instance of the server.
+
+You try it out and realize that the server segfaults when there are no matches.
+
+You despair. There goes your flow. The name of the backend engineers flash
+before your eyes. You consider not inviting them to your wedding. You make a
+mental list of who's most likely to have introduced the bug. You consider
+quitting your job and becoming a barista in a different country -- maybe even
+in New Orleans.  You'll need a specialty, maybe peanut butter cookies. You'll
+kill it with your peanut butter cookies. People will know about them from New
+England to New Mexico. You'll become famous and, in a few years, run for
+president of the United States. Then you consider the logistics: your fiancée
+will need to quit her job, you'll need to hire a moving service, your dog may
+not survive the flight.
+
+You stash your local changes and set out to write a test case reproducing the
+issue.
+</div>
+
+Unless you're a quality analyst, you probably don't go around trying to find
+bugs for fun. Most of us discover bugs while trying to get something done:
+maybe you're simply using the server, or you're writing a new feature that
+depends on a different feature or function. Discovering wrong behavior in a
+codebase is almost always more of an annoyance than a nice surprise.
+
+Had the bug been caught before merge, the bug reporter would most likely have
+been the person who introduced the faulty behavior, and would have most likely
+been expecting to encounter "some" bug (correct software on the first attempt
+is a myth, sorry).
+
+### It creates noise
+
+<div class="story">
+
+> you> @channel I'm experiencing server crashes when there are no matches on
+> `/users/foo/notes?query=bar`, anybody knows anything about this?
+>
+> tom> please use @here instead of @channel
+>
+> pat> @tom please don't use "at" channel
+>
+> jen> @you yes I think this is related to #2551
+>
+> mol> @jen #2551 was fixed last week
+>
+> jen> okay nevermind, no idea @you :(
+>
+> mkt> @channel lunch?
+
+</div>
+
+Whether you are reporting issues on Slack, GitHub, JIRA, you name it, reporting
+a bug creates some noise. Some people will need to label the ticket, will try
+to help you figure out the cause, others may simply get distracted because of
+the extra Slack/GitHub/JIRA notifications.
+
+All this can be avoided if bugs are caught before they are merged to master:
+bugs are typically not reported before the code is part of the mainstream
+codebase.
```

Comment:
not sure about this because you would still need to report the crash, so there would still be _some_ noise?